### DISCUSSION TITLE
Which is preferable: column with Array type or many rows?

### DISCUSSION THREAD

**Discussion by vincenzon** [ClickHouse/ClickHouse#48695] [Q&A] [answered]
I have asked myself this a few times when building out schema, so I thought I would ask here. Which is better (for insert and subsequent query performance) having a column with an array type or broadcasting that column across several rows? 

I understand, the answer is always "it depends". But I am looking for guidance on how to reason about the trade-offs. 

Say I have data with M values of plain data types (numbers and strings), and one additional value that is an array with on-average m values. I have a process that generates these values at a rate of say R / second. I want to store these values in a table. 

One option is to store the values as M columns, and the array value as a single column with an Array type. 

A second option is to store the values as M+1 columns, with each element of the array value stored individually and the M plain data values repeated. 

I assume that the on disk storage in both cases will roughly be the same due to column compression.

I also assume that the insert rate for the first case will be faster, and the query rate will be slower. And the inverse for the second case, slower insert and faster queries.

What I am asking is, in terms of M, m, and R, how much of a difference would I expect to see between the two cases? 

As R increases does one strategy dominate the other?   

I have been assuming that as M increases, the array column approach is better because insert costs increase and the _difference_ in query costs remain the same. Is that correct? 

Similarly, I assume that as m increases, insert costs increase linearly, and query costs decrease, but by how much?

I look forward to hearing comments, and of course if I am overlooking any unknown-unknowns. Thanks,

Matt

**Comment by den-crane** [comment]
https://kb.altinity.com/altinity-kb-schema-design/best-schema-for-storing-many-metrics-registered-from-the-single-source/

**Comment by vincenzon** [comment]
Thanks Denny, that documentation is useful. I think my problem / question is a bit different though. In my case the individual measurements have a small set of scalar metrics and a single array valued metric. The array value is not a collection of separate metrics but rather the type of the metric as naturally observed. For example, a similar measurement for a calendar app might be:

```
meetingId   Int64
date        DateTime
topic       String
duration    Int64
attendees   Array<String>
```

I am debating saving these to a table with the columns typed as given versus storing each measurement as multiple rows (one row for each attendee, column attendee typed as String, with the other columns repeated).

While it would involve bigger insert traffic (unless, can ClickHouse do an array join on insert?), many of the queries against this data would involve forming the array join over attendees and querying the resulting intermediate table, so just creating the table that way seems like it might be more efficient in the long run.

Thanks for the help.

**Comment by den-crane** [comment, marked as answer]
>many of the queries against this data would involve forming the array join over attendees and querying the resulting intermediate table

`array join` is cheap, and streamable, there is no intermediate table

It depends on the queries.

And Clickhouse is able to do ArrayJoin during insertion.

**Comment by vincenzon** [comment]
Interesting, so, I could equalize the insert costs between the two options using an array join on insertion, but I don't have to because array join is cheap anyway! 

I guess having the table match the way the data is generated in the first place makes me lean towards keeping the metric array typed.

For completeness, how would one effect an array join on insert?
